src/Functions/parseDateTime.cpp

```diff
@@ -0,0 +1,1766 @@
+                    while (cur < end && cur < start + max_digits_consume && *cur >= '0' && *cur <= '9')
+                    {
+                        number = number * 10 + (*cur - '0');
+                        ++cur;
+                        ++count;
+                    }
+                    if (count == 2)
+                    {
+                        if (number >= 70)
+                            number += 1900;
+                        else if (number >= 0 && number < 70)
+                            number += 2000;
+                    }
+                    else
+                    {
+                        while (cur < end && cur < start + max_digits_consume && *cur >= '0' && *cur <= '9')
+                        {
+                            number = number * 10 + (*cur - '0');
+                            ++cur;
+                        }
+                    }
+                }
+                else
+                {
+                    while (cur < end && cur < start + max_digits_consume && *cur >= '0' and *cur <= '9')
+                    {
+                        number = number * 10 + (*cur - '0');
+                        ++cur;
+                    }
+                }
+
+                /// Need to have read at least one digit.
+                if (cur <= start)
+                    throw Exception(
+                        ErrorCodes::CANNOT_PARSE_DATETIME,
+                        "Unable to parse flag {} from {} because read number failed",
+                        flag,
+                        std::string_view(cur, end - cur));
+
+                if (negative)
+                    number *= -1;
+
+                return cur;
+            }
+
+            static Pos jodaEra(int, Pos cur, Pos end, [[maybe_unused]] const String & flag, DateTime & date)
+            {
+                checkSpace(cur, end, 2, "jodaEra requires size >= 2", flag);
+
+                String era(cur, 2);
+                date.setEra(era);
+                cur += 2;
+                return cur;
+            }
+
+            static Pos jodaCenturyOfEra(int repetitions, Pos cur, Pos end, [[maybe_unused]] const String & flag, DateTime & date)
+            {
+                Int32 century;
+                cur = readNumberWithVariableLength(cur, end, false, false, false, repetitions, repetitions, flag, century);
+                date.setCentury(century);
+                return cur;
+            }
+
+            static Pos jodaYearOfEra(int repetitions, Pos cur, Pos end, [[maybe_unused]] const String & flag, DateTime & date)
+            {
+                Int32 year_of_era;
+                cur = readNumberWithVariableLength(cur, end, false, false, true, repetitions, repetitions, flag, year_of_era);
+                date.setYear(year_of_era, true);
+                return cur;
+            }
+
+            static Pos jodaWeekYear(int repetitions, Pos cur, Pos end, [[maybe_unused]] const String & flag, DateTime & date)
```

Comment:
>  But we reserve only space for at most 4 repetitions.
Sorry, I do not understand where the issue is?  Since `reserve_size += std::max(repetitions, 4);` and `repetitions == 8`, it seems all right.

Reply:
If I see correctly,
- `reserve_size += std::max(repetitions, 4);` adds \+4 characters to the template but
- `jodaWeekYear` (specifically `writeNumberWithPadding`) writes 8 characters into the output.